sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala

```diff
@@ -139,6 +139,14 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
       ("a", 1), ("b", 2))
   }
 
+  test("as tuple using instance encoder") {
+    val data = Seq(("a", 1), ("b", 2), ("c", 3))
+    val ds = data.toDS()
+    checkDataset(
+      ds.toDF().as[(String, Int)](ds.encoder),
```

Comment:
This example is confusing. Without explicitly specifying encoder, we still can convert it to Dataset. Could you give a proper example?

Reply:
I am not sure what you mean by a proper example. Given the change in the PR, the only purpose of the test is to show (a) that the `encoder` property can be accessed and (b) that the accessed encoder will work correctly for another dataset instance. In that context, the fact that conversion without providing an explicit encoder is possible is not relevant. Further, adding a more complicated example where implicit conversion is not possible will add nothing to the purpose of the test.

Reply:
@ssimeonov . What @gatorsmile meant is that the following test works without your patch. So, the question is why do we need your patch? Please provide a valid test case instead of just a simple `ds.encoder` usage.
```scala
  test("as tuple using instance encoder") {
    val data = Seq(("a", 1), ("b", 2), ("c", 3))
    val ds = data.toDS()
    checkDataset(
      ds.toDF().as[(String, Int)],
      data: _*)
  }
```

Reply:
@dongjoon-hyun I am not sure what your are trying to suggest with your test example other than prove my point, which is that the test I added is the minimum test that verifies the added functionality: making `encoder` publicly accessible.

The purpose of test cases is not to motivate functionality but to test functionality. The motivation is provided in the JIRA issue. I believe that explanation to be perfectly clear or else @gatorsmile wouldn't have asked me to create a PR. 

I do not understand the point of creating a large & complicated test that *requires* a public `encoder`. As outlined in the JIRA issue, that test would require:

1. A framework interface including something like `def foo[A](ds: Dataset[A]): Dataset[A]`
2. An implementation of that interface
3. Some complex and rather arbitrary ad hoc data transformations in the implementation 

Do you really think this test would be easy to comprehend amidst the simple tests in the suite? Do you really think the test would add any meaningful verification of the tiny code change in patch?

Reply:
If you think so, shall we remove this? It looks misleading to me, too.

Reply:
This is one of the simplest contrived examples I can come up with. Is that what you are looking for, used in a test?

```scala
def firstRowByGroup[A](groupBy: Column)(ds: Dataset[A]): Dataset[A] =
  ds.groupBy(groupBy).agg(first(struct('*), ignoreNulls = true).as("row")).select("row.*").as[A](ds.encoder)
```

Reply:
I would be OK with removing it, yes. IMO no test is cleaner than even a test based on something like the above function.

Reply:
I think the change is very straightforward: we want to make a field public. Maybe we can just remove this test

Reply:
Perfect, will do.

Reply:
Done